gax/src/main/java/com/google/api/gax/retrying/StreamTracker.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2017, Google LLC All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.retrying;
+
+import com.google.api.core.BetaApi;
+
+/**
+ * This is part of the server streaming retry api. It's implementors are responsible for tracking
+ * the progress of the stream and calculating a request to resume it in case of an error.
+ *
+ * <p>Implementations don't have to be threadsafe because all of the calls will be serialized.
+ */
+@BetaApi("The surface for streaming is not stable yet and may change in the future.")
+public interface StreamTracker<RequestT, ResponseT> {
+  /**
+   * Called by the {@link RetryingServerStream} to notify of a successfully received response.
+   *
+   * @param response
+   */
+  void onProgress(ResponseT response);
+
+  /**
+   * Called when a stream needs to be restarted, the implementation should generate a request that
+   * will yield a stream whose first response would come right after the last response in
+   * onProgress.
+   */
+  RequestT getResumeRequest(RequestT originalRequest);
+
+  /**
+   * Stateless factory to create new {@link StreamTracker}s to use new calls in
+   * RetryingServerStreamingCallable.
+   */
+  interface Factory<RequestT, ResponseT> {
+    StreamTracker<RequestT, ResponseT> create();
+  }
```

Comment:
I think I would like to avoid an inner entity here, and go even further to avoid this Factory altogether. My suggestion is to add a "createNew" method on StreamTracker to create a new instance of the same implementing type.